pkg/executor/aggregate/agg_spill.go

```diff
@@ -15,15 +15,282 @@
+		cond                       *sync.Cond
+		nextPartitionIdx           int
+		spilledChunksIO            [][]*chunk.DataInDiskByChunks
+		spillTriggered             int32
+		isSpilling                 bool
+		isAllPartialWorkerFinished bool
+	}
+
+	memTracker     *memory.Tracker
+	diskTracker    *disk.Tracker
+	isPartialStage int32
+	hasError       int32
+
+	// These agg functions are partial agg functions that are same with partial workers'.
+	// They only be used for restoring data that are spilled to disk in partial stage.
+	aggFuncsForRestoring []aggfuncs.AggFunc
+
+	// When spill is triggered, all partial workers should stop
+	// and we can ensure this when the spill action gets write lock.
+	syncLock sync.RWMutex
+}
+
+func newSpillHelper(tracker *memory.Tracker, aggFuncsForRestoring []aggfuncs.AggFunc) *parallelHashAggSpillHelper {
+	mu := new(sync.Mutex)
+	return &parallelHashAggSpillHelper{
+		lock: struct {
+			mu                         *sync.Mutex
+			cond                       *sync.Cond
+			nextPartitionIdx           int
+			spilledChunksIO            [][]*chunk.DataInDiskByChunks
+			spillTriggered             int32
+			isSpilling                 bool
+			isAllPartialWorkerFinished bool
+		}{
+			mu:                         mu,
+			cond:                       sync.NewCond(mu),
+			spilledChunksIO:            make([][]*chunk.DataInDiskByChunks, spilledPartitionNum),
+			spillTriggered:             0,
+			isSpilling:                 false,
+			nextPartitionIdx:           spilledPartitionNum - 1,
+			isAllPartialWorkerFinished: false,
+		},
+		memTracker:           tracker,
+		isPartialStage:       partialStageFlag,
+		hasError:             0,
+		aggFuncsForRestoring: aggFuncsForRestoring,
+	}
+}
+
+func (p *parallelHashAggSpillHelper) close() {
+	for _, ios := range p.lock.spilledChunksIO {
+		for _, io := range ios {
+			io.Close()
+		}
+	}
+}
+
+func (p *parallelHashAggSpillHelper) getNextPartition() (int, bool) {
+	p.lock.mu.Lock()
+	defer p.lock.mu.Unlock()
+	partitionIdx := p.lock.nextPartitionIdx
+	if partitionIdx < 0 {
+		return -1, false
+	}
+
+	p.lock.nextPartitionIdx--
+	return partitionIdx, true
+}
+
+func (p *parallelHashAggSpillHelper) addListInDisks(dataInDisk []*chunk.DataInDiskByChunks) {
+	p.lock.mu.Lock()
+	defer p.lock.mu.Unlock()
+	for i, data := range dataInDisk {
+		p.lock.spilledChunksIO[i] = append(p.lock.spilledChunksIO[i], data)
+	}
+}
+
+func (p *parallelHashAggSpillHelper) getListInDisks(partitionNum int) []*chunk.DataInDiskByChunks {
+	p.lock.mu.Lock()
+	defer p.lock.mu.Unlock()
+	return p.lock.spilledChunksIO[partitionNum]
+}
+
+func (p *parallelHashAggSpillHelper) isSpillTriggered() bool {
+	p.lock.mu.Lock()
+	defer p.lock.mu.Unlock()
+	return p.lock.spillTriggered == spillFlag
+}
+
+func (p *parallelHashAggSpillHelper) setSpillTriggered() {
+	p.lock.mu.Lock()
+	defer p.lock.mu.Unlock()
+	p.lock.spillTriggered = spillFlag
+}
+
+func (p *parallelHashAggSpillHelper) isInPartialStage() bool {
```

Comment:
I don't see the status change for `isPartialStage`, looks like it is always in partial stage?

Reply:
> I don't see the status change for `isPartialStage`, looks like it is always in partial stage?

yes, and I think this `p.isPartialStage` is useless as we will call `checkAllPartialWorkersFinished`.